File: leet/Arrays/Java/FindMinInRotatedSortedArray.java
import java.util.Arrays;

// I: array
// O: array
// EDGE CASE: array has one item
// CONSTRAINTS: O(log(n))

// time complexity: O(log(n))
// space complexity:  

// Objective: return an sorted array from min to max by "rotating"
// given array
// hint: use binary search

public class FindMinInRotatedSortedArray {
  public static void main(String[] args) {
    int[] nums = new int[] {2, 3, 4, 5, 1};
    int[] nums2 = new int[] {4,5,6,7,0,1, 2};
    System.out.println("nums " + findMin(nums));
    System.out.println("nums2 " + findMin(nums2));
  }

  public static int findMin(int[] nums) {
    if(nums==null || nums.length==0)
    return -1;

int i=0; 
int j=nums.length-1;

while(i<=j){
  System.out.println(nums[i] + " " + nums[j]);
    if(nums[i]<=nums[j])
        return nums[i];

    int m=(i+j)/2;

    if(nums[m]>=nums[i]){
        i=m+1;
    }else{
        j=m;
    }
}

return -1;
}
}

File: leet/Arrays/Javascript/find-min -in-rotated-sorted-array.js
// I: array
// O: array
// EDGE CASE: array has one item
// CONSTRAINTS: O(log(n))

// time complexity: O(log(n))
// space complexity:  

// Objective: return an sorted array from min to max by "rotating"
// given array
// hint: use binary search and recursion

function findMin(nums) {
  // for outlier condition
  if (nums.length === 1) return nums

  // find middle index of array
  const midIndex = Math.floor(nums.length / 2)
  
  // if value at center of array is less than value
  // to the left of the center return middle index value
  // ends recursion
  if(nums[midIndex] < nums[midIndex - 1]) {
    return nums[midIndex]
  // if value at center of array is greater than value
  // to the left of the center recurse array using values
  // to the right of the middle index
  } else if (nums[midIndex] > nums[nums.length - 1]) { 
    return findMin(nums.slice(midIndex + 1))
  }
  // recursive call to find min using the values
  // to the left of middle index value an include middle index value
  return findMin(nums.slice(0, midIndex))
}



let nums = [2, 3, 1]
let nums1 = [3,4,5,1,2]
let nums2 = [4,5,6,7,0,1,2]
let nums3 = [11,13,15,17]
console.log(findMin(nums))
// console.log(findMin(nums1))
// console.log(findMin(nums2))
// console.log(findMin(nums3))